gensim/models/fasttext_corpusfile.pyx

```diff
@@ -0,0 +1,287 @@
+
+    cdef CythonLineSentence input_stream = CythonLineSentence(corpus_file, offset)
+    cdef CythonVocab vocab = _cython_vocab
+
+    cdef int i, j, k
+    cdef int effective_words = 0, effective_sentences = 0
+    cdef int total_effective_words = 0, total_sentences = 0, total_words = 0
+    cdef int sent_idx, idx_start, idx_end
+
+    init_ft_config(&c, model, _alpha, _work, _neu1)
+
+    # for preparing batches & training
+    cdef vector[vector[string]] sentences
+    cdef unsigned long long random_number
+    cdef VocabItem word
+
+    with nogil:
+        input_stream.reset()
+        while not (input_stream.is_eof() or total_words > expected_words / c.workers):
+            effective_sentences = 0
+            effective_words = 0
+
+            sentences = input_stream.next_batch()
+
+            prepare_c_structures_for_batch(sentences, c.sample, c.hs, c.window, &total_words, &effective_words,
+                                           &effective_sentences, &c.next_random, vocab.get_vocab_ptr(), c.sentence_idx,
+                                           c.indexes, c.codelens, c.codes, c.points, c.reduced_windows,
+                                           c.subwords_idx_len, c.subwords_idx)
+
+            for sent_idx in range(effective_sentences):
+                idx_start = c.sentence_idx[sent_idx]
+                idx_end = c.sentence_idx[sent_idx + 1]
+                for i in range(idx_start, idx_end):
+                    j = i - c.window + c.reduced_windows[i]
+                    if j < idx_start:
+                        j = idx_start
+                    k = i + c.window + 1 - c.reduced_windows[i]
+                    if k > idx_end:
+                        k = idx_end
+
+                    if c.hs:
+                        fasttext_fast_sentence_cbow_hs(
+                            c.points[i], c.codes[i], c.codelens, c.neu1, c.syn0_vocab, c.syn0_ngrams, c.syn1, c.size,
+                            c.indexes, c.subwords_idx, c.subwords_idx_len, c.alpha, c.work, i, j, k, c.cbow_mean,
+                            c.word_locks_vocab, c.word_locks_ngrams)
+                    if c.negative:
+                        c.next_random = fasttext_fast_sentence_cbow_neg(
+                            c.negative, c.cum_table, c.cum_table_len, c.codelens, c.neu1, c.syn0_vocab, c.syn0_ngrams,
+                            c.syn1neg, c.size, c.indexes, c.subwords_idx, c.subwords_idx_len, c.alpha, c.work, i, j, k,
+                            c.cbow_mean, c.next_random, c.word_locks_vocab, c.word_locks_ngrams)
+
+            total_sentences += sentences.size()
+            total_effective_words += effective_words
+
+            c.alpha = get_next_alpha(start_alpha, end_alpha, total_sentences, total_words,
+                                     expected_examples, expected_words, cur_epoch, num_epochs)
+
+    return total_sentences, total_effective_words, total_words
+
+
+CORPUSFILE_VERSION = 1
```

Comment:
nitpick: missing `\n`  (here and everywhere)

Reply:
done